pkg/apis/rollouts/v1alpha1/analysis_types.go

```diff
@@ -106,6 +106,8 @@ type Metric struct {
 	ConsecutiveErrorLimit *intstrutil.IntOrString `json:"consecutiveErrorLimit,omitempty" protobuf:"bytes,9,opt,name=consecutiveErrorLimit"`
 	// Provider configuration to the external system to use to verify the analysis
 	Provider MetricProvider `json:"provider" protobuf:"bytes,10,opt,name=provider"`
+	// Whether to evaluate this metric in a Dry-Run mode
+	DryRun bool `json:"dryRun,omitempty" protobuf:"varint,11,opt,name=dryRun"`
```

Comment:
Do we anticipate users will want to parameterize this? If so we may need a different datatype than `bool` since bools cannot be string replaced from arguments.

Alternatively, we might want to rethink the syntax to perform dry runs. For example, it could be at the spec level:

Dryrun sucess-rate and latency, but not saturation.
```yaml
spec:
  metrics:
  - name: success-rate
    interval: 10m
...
  - name: latency
    interval: 10m
...
  - name: saturation
    interval: 10m
...
  dryRun: "success-rate,latency"
```

Dry run everything:
```yaml
spec:
  metrics:
  - name: success-rate
    interval: 10m
...
  - name: latency
    interval: 10m
  dryRun: "*"
```

And we could improve the handling such that `dryRun` could be replaced with arguments.

Reply:
@jessesuen I was thinking of it this way:

**1).** It could be done at metric level like this:
```json
metrics:
  - name: total-5xx-errors
    interval: 5m
    dryRun: true
    failureCondition: result[0] >= 10
    failureLimit: 3
    provider:
      ...
```

**2).** It could be done at the **AnalysisRun** level like this:
```json
spec:
  metrics:
  - name: success-rate
    interval: 10m
...
  - name: latency
    interval: 10m
  dryRun: "*"
```

```json
spec:
  metrics:
  - name: success-rate
    interval: 10m
...
  - name: latency
    interval: 10m
  dryRun: "success-rate, latency"
```

In this PR, I have implemented **Case 1**. I'll create a follow-up PR with docs etc. for **Case 2**.

What do you think?

Reply:
After thinking about this more, I think a spec level AnalysisRun setting is the more appropriate place for dry-run decisions because tightly coupling with the metric does not allow flexibility for different rollouts/users to decide if they want to opt-in to the metric. If we were to have it at the metric level, it makes the decision part of the Analysis**_Template_** instead of the Analysis**_Run_**. For example, with ClusterAnalysisTemplates, an operator should not unilaterally decide to make a metric dry-run for all rollouts in the cluster. This is a decision that should be reserved to Rollout users.

But I have one improvement to the syntax which I think will make it more explicit and canonically kubernetes, which is to use a list instead of a comma-delimited string.

Dry-Run success-rate and latency but not saturation:

```yaml
spec:
  metrics:
  - name: success-rate
    interval: 10m
...
  - name: latency
    interval: 10m
...
  - name: saturation
    interval: 10m
...
  dryRun:
  - metricName: success-rate
  - metricName: latency
```

The list would support wildcards if they want to dry run everything:

```yaml
  dryRun:
  - metricName: *
```

Also notice that the dryRun list is a list of structured objects instead of primitives. I suggest this because it may allow us to expand on the dryRun capabilities in the future:

```yaml
  dryRun:
  - metricName: success-rate
    futureFeature: true
```

When a Rollout wants to dry run their analysis, they simply need to specify the dryRun flag to the analysis stanza. e.g.:

```yaml
      - analysis:
          templates:
          - templateName: random-fail
          - templateName: always-pass
          dryRun:
          - metricName: *
```

Which we would simply/blindly carry over when constructing the AnalysisRun resource.